### ISSUE TITLE
Loading a xgboost model with DMatrix will crash python and R

### ISSUE THREAD

**Issue by qinwf** [dmlc/xgboost#2849]
For openMP version of xgboost,

If the program accidentally load a xgboost model with DMatrix, it will crash python or R.

## Environment info

Operating System: Ubuntu 16.04

Compiler:  GCC 5.4.0

Package used (python/R/jvm/C++): python, R

`xgboost` version used: 

+ 0.6a2 from pip3

+ R package from http://dmlc.ml/drat/

## Steps to reproduce

1. create any xgboost model file or download a model file (and unzip it to iris.xg.model)

[iris.xg.model.zip](https://github.com/dmlc/xgboost/files/1430262/iris.xg.model.zip)

```
#!/usr/bin/python3
import xgboost as xgb
from sklearn import datasets

iris = datasets.load_iris()
dtrain = xgb.DMatrix(iris.data, label = iris.target)
param = {
    'max_depth': 3,
    'eta': 0.3,
    'objective': 'multi:softprob',
    'num_class': 3
}
num_round = 20
bst = xgb.train(param, dtrain, num_round)
bst.save_model("iris.xg.model")
```

2.load the model file with DMatrix

For Python:

```python3
xgb.DMatrix('iris.xg.model')

[06:09:23] dmlc-core/include/dmlc/logging.h:235: [06:09:23] src/data/././strtonum.h:141: Check failed: (sign) == (true)
terminate called after throwing an instance of 'dmlc::Error'
  what():  [06:09:23] src/data/././strtonum.h:141: Check failed: (sign) == (true)
[1]    26684 abort (core dumped)  python3
```

For R:

```R
xgboost::xgb.DMatrix('iris.xg.model')

terminate called recursively
[1]    27747 abort (core dumped)  R
```

## What have you tried?

1. reproduce this error
2. this error is thrown from OpenMP code, so it cannot be caught by the API_END

**Comment by airkid** [comment]
I got the same error while running basic_walkthrough.py in /demo/
`python get_started.py                                                 134 ↵
[[[[17:04:3017:04:3017:04:3017:04:30] ] ] ] dmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.h::::300300300300: : : : [17:04:30] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1) 

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000011379ceb9 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x0000000113b83fa0 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[17:04:30] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1) 

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000011379ceb9 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x0000000113b83fa0 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[17:04:30] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1) 

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000011379ceb9 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x0000000113b83fa0 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[17:04:30] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1) 

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000011379ceb9 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x0000000113b83fa0 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16`

**Comment by tk3369** [comment]
Bump

**Comment by hcho3** [comment]
I see how crashing the program is undesirable. The issue is that in thread-enabled XGBoost, there are actually multiple threads parsing the input file. So when the parser encounters an invalid byte, there is currently no good way to catch the exception other than calling `std::terminate()` (which terminates the whole program). Any suggestion for catching exceptions thrown by threads?

Note. This behavior applies to both custom threads and OpenMP threads. OpenMP specifically "forbids" code that lets exception escape the parallel region.

**Comment by sunils27** [comment]
+1, the GPU version of the xgboost library is almost unusable for me at the moment

**Comment by berli** [comment]
+1 same here

**Comment by hcho3** [comment]
@sunils27 @berli @mglomba @tk3369 @airkid 
The issue appears to be fixed by PR #3221. The PyPI package does not have this fix yet, so for now please compile from the latest source.
